terraso_backend/apps/story_map/views.py

```diff
@@ -0,0 +1,213 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see https://www.gnu.org/licenses/.
+
+import json
+import uuid
+from dataclasses import asdict
+from datetime import datetime
+
+import rules
+import structlog
+from django.core.exceptions import NON_FIELD_ERRORS, ValidationError
+from django.db import IntegrityError
+from django.http import JsonResponse
+from django.utils import timezone
+from django.views.generic.edit import FormView
+
+from apps.auth.mixins import AuthenticationRequiredMixin
+from apps.core.exceptions import ErrorContext, ErrorMessage
+
+from .forms import StoryMapForm
+from .models import StoryMap
+from .services import story_map_media_upload_service
+
+logger = structlog.get_logger(__name__)
+
+
+class StoryMapAddView(AuthenticationRequiredMixin, FormView):
+    def post(self, request, **kwargs):
+        form_data = request.POST.copy()
+
+        config = json.loads(form_data["configuration"])
+
+        form_data["created_by"] = request.user
+        form_data["is_published"] = form_data["is_published"] == "true"
+        form_data["published_at"] = (
+            timezone.make_aware(datetime.now(), timezone.get_current_timezone())
+            if form_data["is_published"]
+            else None
+        )
+
+        entry_form = StoryMapForm(data=form_data)
+
+        if not entry_form.is_valid():
+            error_messages = get_error_messages(entry_form.errors.as_data())
+            return JsonResponse(
+                {"errors": [{"message": [asdict(e) for e in error_messages]}]}, status=400
+            )
+
+        try:
+            story_map = StoryMap.objects.create(
+                created_by=form_data["created_by"],
+                title=form_data["title"],
+                is_published=form_data["is_published"],
+                configuration=handle_config_media(config, None, request),
+                published_at=form_data["published_at"],
+            )
+        except IntegrityError as exc:
+            return handle_integrity_error(exc)
+
+        return JsonResponse(story_map.to_dict(), status=201)
+
+
+class StoryMapUpdateView(AuthenticationRequiredMixin, FormView):
+    def post(self, request, **kwargs):
+        user = request.user
+        form_data = request.POST.copy()
+
+        story_map = StoryMap.objects.get(id=form_data["id"])
+
+        if not rules.test_rule("allowed_to_change_story_map", user, story_map):
+            logger.info(
+                "Attempt to update a StoryMap, but user lacks permission",
+                extra={"user_id": user.pk, "story_map_id": str(story_map.id)},
+            )
+            error_message = ErrorMessage(
+                code="update", context=ErrorContext(model="StoryMap", field=NON_FIELD_ERRORS)
+            )
+            return JsonResponse({"errors": [{"message": [asdict(error_message)]}]}, status=400)
```

Comment:
q: is 400 the appropriate error code here?

Reply:
Yes, this is a validation error for the data the user is sending so 400 - Bad request makes sense